server/src/main/java/org/elasticsearch/index/fielddata/MultiValuedSortedBinaryDocValues.java

```diff
@@ -54,10 +59,88 @@ public int docValueCount() {
     }
 
     @Override
-    public BytesRef nextValue() throws IOException {
-        scratch.length = in.readVInt();
-        scratch.offset = in.getPosition();
-        in.setPosition(scratch.offset + scratch.length);
-        return scratch;
+    public abstract boolean advanceExact(int doc) throws IOException;
+
+    @Override
+    public abstract BytesRef nextValue() throws IOException;
+
+    /**
+     * Multivalued binary doc values encoded by {@link org.elasticsearch.index.mapper.BinaryFieldMapper.CustomBinaryDocValuesField}.
+     * These have the form: [doc value count][length of value 1][value 1][length of value 2][value 2]...
+     */
+    private static class IntegratedCounts extends MultiValuedSortedBinaryDocValues {
+        IntegratedCounts(BinaryDocValues values) {
+            super(values);
+        }
+
+        @Override
+        public boolean advanceExact(int doc) throws IOException {
+            if (values.advanceExact(doc)) {
+                final BytesRef bytes = values.binaryValue();
+                assert bytes.length > 0;
+                in.reset(bytes.bytes, bytes.offset, bytes.length);
+                count = in.readVInt();
+                scratch.bytes = bytes.bytes;
+                return true;
+            } else {
+                count = 0;
+                return false;
+            }
+        }
+
+        @Override
+        public BytesRef nextValue() throws IOException {
+            scratch.length = in.readVInt();
+            scratch.offset = in.getPosition();
+            in.setPosition(scratch.offset + scratch.length);
+            return scratch;
+        }
+    }
+
+    /**
+     * Multivalued binary doc values with counts stored in a separate numeric doc values.
+     * If a binary value contains a single value, payload is of the form: [value 1]
+     * If a binary value contains multiple values, payload is of the form: [length of value 1][value 1][length of value 2][value 2]...
+     */
+    private static class SeparateCounts extends MultiValuedSortedBinaryDocValues {
+        private final NumericDocValues counts;
+
+        SeparateCounts(BinaryDocValues values, NumericDocValues counts) {
+            super(values);
+            this.counts = counts;
+        }
+
+        @Override
+        public boolean advanceExact(int doc) throws IOException {
+            if (values.advanceExact(doc)) {
+                assert counts.advanceExact(doc);
+                count = Math.toIntExact(counts.longValue());
+
+                final BytesRef bytes = values.binaryValue();
+
+                if (count == 1) {
+                    scratch.bytes = bytes.bytes;
+                    scratch.offset = bytes.offset;
+                    scratch.length = bytes.length;
+                } else {
+                    in.reset(bytes.bytes, bytes.offset, bytes.length);
+                    scratch.bytes = bytes.bytes;
+                }
```

Comment:
I think we can post pone this now and do this when we really need to in `nextValue()`?